### ISSUE TITLE
training diverges/loss increasing

### ISSUE THREAD

**Issue by zcyang** [cvxgrp/cvxpylayers#69]
Hi,

I was testing a simple problem and find the training diverges and loss is increasing.

```python

import numpy as np
import torch
import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.autograd import Variable
from torch.nn.parameter import Parameter

import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer

N = 200
M = 100
BOUND = 0.1
NORM_BOUND = 10
NUM_STEPS = 3

class Opt(nn.Module):
  def __init__(self, weight=5.0):
    super(Opt, self).__init__()
    self.weight = Parameter(
      torch.tensor([[weight]], requires_grad=True))

    self.L = torch.randn(N, M)
    L_diag = torch.rand(N)
    L_diag[:N//2] *= 0.1
    L_diag[N//2:] *= 10
    self.L_diag = torch.diag(L_diag)

    _p = cp.Parameter((N))
    _prev = cp.Parameter((N))
    _v = cp.Variable((N))
    obj = cp.Minimize(0.5*cp.sum_squares(self.L.T@_v)
                      + 0.5*cp.sum_squares(self.L_diag@_v)
                      + 50*cp.sum_squares(_v - _prev)
                      - _v@_p)
    cons = [_v <= BOUND,
            -_v <= BOUND,
            cp.norm(_v, 1) <= NORM_BOUND
    ]
    prob = cp.Problem(obj, cons)
    self.cvxopt = CvxpyLayer(prob, [_p, _prev], [_v])

  def forward(self, x):
    num_steps = x.shape[0]
    xw = torch.matmul(x, self.weight).squeeze()
    init_v = torch.zeros(x.shape[1])
    v_list = []
    for i in range(num_steps):
      v, = self.cvxopt(xw[i], init_v, solver_args={"eps": 1e-7, "max_iters": 200000})
      init_v = v
      v_list.append(v)
    v = torch.stack(v_list)

    return v

def main():
  torch.manual_seed(0)
  opt = Opt(weight=5.0)

  x = torch.randn(N, 1)
  x = x.unsqueeze(0).repeat(NUM_STEPS, 1, 1)
  y = 0.1*x[-1] + torch.randn(N, 1)

  optimizer = optim.Adam(opt.parameters(), 1e-1)
  for i in range(200000):
    v = opt(x)
    loss = -torch.sum(y*v[-1])
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    l = loss.detach().numpy().tolist()
    w = opt.weight.detach().numpy()
    g = opt.weight.grad.detach().numpy()
    print("loss {:.3f} weight {:.3f} "
          "grad {:.3f} ".format(
            l, w[0][0], g[0][0],
          ))


if __name__ == "__main__":
  main()


``` 

The training log looks like this:

```
loss -8.689 weight 4.900 grad 0.564
loss -8.668 weight 4.800 grad 0.508
loss -8.638 weight 4.701 grad 0.475
loss -8.603 weight 4.603 grad 0.405
loss -8.563 weight 4.506 grad 0.398
loss -8.520 weight 4.411 grad 0.292
loss -8.470 weight 4.321 grad 0.196
loss -8.415 weight 4.232 grad 0.283
loss -8.344 weight 4.163 grad -0.213
loss -8.281 weight 4.104 grad -0.079
loss -8.233 weight 4.056 grad -0.078
loss -8.192 weight 4.017 grad -0.120
loss -8.162 weight 3.985 grad -0.081
loss -8.137 weight 3.959 grad -0.056
...
```

Also used the finite difference method to test the gradient at 5.0 following #67 
Got following output (by setting solver_args={"eps": 1e-8, "max_iters": 200000}:

grad with backward:
```
[[0.5678954]]
```
grad with finite difference:
```
tensor(-0.0734, grad_fn=<DivBackward0>)
```
, which has different signs.

Also tried to set everything as `torch.double` and set `eps=1e-9`, neither solves the problem.

**Comment by SteveDiamond** [comment]
@sbarratt is the ECOS solver gone?

**Comment by sbarratt** [comment]
It’s not merged yet

**Comment by zcyang** [comment]
> @zcyang there is preliminary support for the ECOS solver here: [cvxgrp/diffcp#37](https://github.com/cvxgrp/diffcp/pull/37)
> but it hasn't been merged into diffcp yet (which is called by cvxpylayers to solve the convex problem). The ECOS solver is higher accuracy and might give different results.

Does the accuracy of the gradient purely depends on the optimization solution accuracy? 
Or computing the gradient itself (even given a perfect solution) has some level of noise that depends o the solver?

If it is the second case, then changing the solver may help?

If it is the first case, I have adapted the code to output the difference in solution between SCS and EOCS below. It seems the difference in solution is very small. 

```python
import numpy as np
import torch
# torch.set_default_dtype(torch.double)

import torch.nn as nn
import torch.optim as optim
import torch.nn.functional as F
from torch.autograd import Variable
from torch.nn.parameter import Parameter

import cvxpy as cp
from cvxpylayers.torch import CvxpyLayer

N = 200
M = 100
BOUND = 0.1
NORM_BOUND = 10
NUM_STEPS = 3

class Opt(nn.Module):
  def __init__(self, weight=5.0):
    super(Opt, self).__init__()
    self.weight = Parameter(
      torch.tensor([[weight]], dtype=torch.double, requires_grad=True))
    self.L = torch.randn(N, M)
    L_diag = torch.rand(N)
    L_diag[:N//2] *= 0.1
    L_diag[N//2:] *= 10
    self.L_diag = torch.diag(L_diag)

    self.L = self.L.double()
    self.L_diag = self.L_diag.double()

  def forward(self, x):
    num_steps = x.shape[0]
    xw = torch.matmul(x, self.weight).squeeze()
    init_v = torch.zeros(x.shape[1])
    init_v = init_v.double()
    v_list = []
    for i in range(num_steps):
      _p = cp.Parameter((N))
      _prev = cp.Parameter((N))
      _v = cp.Variable((N))
      obj = cp.Minimize(0.5*cp.sum_squares(self.L.T@_v)
                        + 0.5*cp.sum_squares(self.L_diag@_v)
                        + 50*cp.sum_squares(_v - _prev)
                         - _v@_p)
      cons = [_v <= 0.1,
              -_v <= 0.1,
              cp.norm(_v, 1) <= 10
      ]
      prob = cp.Problem(obj, cons)

      _prev.value = init_v.detach().numpy()
      _p.value = xw[i].detach().numpy()

      prob.solve(solver=cp.ECOS)
      v_ecos = _v.value
      if prob.status != "optimal":
        print(prob.status)

      self.cvxopt = CvxpyLayer(prob, [_p, _prev], [_v])

      v, = self.cvxopt(xw[i], init_v) #, solver_args={"eps": 1e-9, "max_iters": 200000})

      print("ecos gap abs {:.6f} ".format(
        np.max(np.abs(v_ecos - v.detach().numpy())),
        ))

      init_v = v
      v_list.append(v)
    v = torch.stack(v_list)

    return v

def main():
  torch.manual_seed(0)
  opt = Opt(weight=5.0)

  x = torch.randn(N, 1)
  x = x.unsqueeze(0).repeat(NUM_STEPS, 1, 1)
  y = 0.1*x[-1] + torch.randn(N, 1)

  x, y = x.double(), y.double()

  optimizer = optim.Adam(opt.parameters(), 1e-1)
  for i in range(200000):
    v = opt(x)
    loss = -torch.sum(y*v[-1])
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    l = loss.detach().numpy().tolist()
    w = opt.weight.detach().numpy()
    g = opt.weight.grad.detach().numpy()
    print("loss {:.3f} weight {:.3f} "
          "grad {:.3f} ".format(
            l, w[0][0], g[0][0],
          ))


``` 

Here is the output log:

```
ecos gap abs 0.000002
ecos gap abs 0.000001
ecos gap abs 0.000001
loss -8.689 weight 4.900 grad 0.565
ecos gap abs 0.000001
ecos gap abs 0.000001
ecos gap abs 0.000000
loss -8.668 weight 4.800 grad 0.503
ecos gap abs 0.000001
ecos gap abs 0.000001
ecos gap abs 0.000001
loss -8.638 weight 4.701 grad 0.487
ecos gap abs 0.000001
ecos gap abs 0.000001
ecos gap abs 0.000001
loss -8.603 weight 4.603 grad 0.407
ecos gap abs 0.000001
ecos gap abs 0.000001
ecos gap abs 0.000001

```
The maximum gap is 1e-6, which is very small.

Do you think changing the solver will make a difference? thanks!

**Comment by zcyang** [comment]
Below is the result using ECOS solver from cvxgrp/diffcp#37:

```
loss -8.689 weight 4.900 grad 0.567
loss -8.668 weight 4.801 grad 0.500
loss -8.638 weight 4.701 grad 0.482
loss -8.603 weight 4.603 grad 0.410
loss -8.563 weight 4.506 grad 0.400
loss -8.520 weight 4.411 grad 0.287
loss -8.470 weight 4.321 grad 0.198
loss -8.415 weight 4.232 grad 0.272
loss -8.345 weight 4.164 grad -0.229
```

The loss is still increasing.

So I guess there must be some problem in the backward (gradient calculation step)? This is also verified in #70 , in which the forward solution from the optimization is the same, while the gradient is inconsistent with sum of losses.

I am not an expert in optimization, but I read the paper a little bit, it seems the backward step involves solving a linear system. Is there any way to control the precision at that step?

I am happy to help to fix this problem, but would like to get your comment and feedback. Thanks!

@SteveDiamond @sbarratt @akshayka

**Comment by bamos** [comment]
There's possibility this is coming from an inaccurate LSQR solve in the backward pass. Can you try passing `mode="dense"`  into `diffcp` (https://github.com/cvxgrp/diffcp/blob/master/diffcp/cone_program.py#L32) by setting `solver_args['mode']='dense'` in the `CvxpyLayer` call here? Same in #70

**Comment by zcyang** [comment]
Thanks @bamos for the suggestion.
Setting `solver_args['mode'] = 'dense'` solves the problem:

```
loss -8.689 weight 5.100 grad -0.070
loss -8.696 weight 5.200 grad -0.070
loss -8.694 weight 5.213 grad 0.091
loss -8.693 weight 5.189 grad 0.091
loss -8.695 weight 5.145 grad 0.091
loss -8.697 weight 5.111 grad -0.013
loss -8.697 weight 5.100 grad -0.070
loss -8.696 weight 5.108 grad -0.070
loss -8.696 weight 5.130 grad -0.070
loss -8.697 weight 5.149 grad 0.002
```

so there must be some problem with the default LSQR solver. This is the same for #70 

But `dense` is several orders of magnitude slower than `lsqr`, especially when `N` is large.

**Comment by bamos** [comment]
It's possible the iteration limit or default thresholds with LSQR there aren't the best for your problem. Debugging this isn't that straightforward because the LSQR code there is in C++ but you can try printing out the iterates here (https://github.com/cvxgrp/diffcp/blob/master/cpp/src/lsqr.cpp#L43), and also passing in other values for the iteration limit/thresholds here: https://github.com/cvxgrp/diffcp/blob/master/diffcp/cone_program.py#L350

Maybe we should add a warning when the LSQR solve is inaccurate? Or tweak some of the defaults here

**Comment by zcyang** [comment]
It turns out that LSQR reaches the iteration limit for this problem.

I changed the mechanism a little bit, by doubling the number of iterations if the `istop == 7`.
https://github.com/cvxgrp/diffcp/blob/master/diffcp/cone_program.py#L350

```python
            r = _diffcp.lsqr(MT, dz)
            iter_lim = 2
            while r.istop == 7:
                print("lsqr reaches iter limit")
                iter_lim *= 2
                r = _diffcp.lsqr(MT, dz, iter_lim=iter_lim)
            r = r.solution
```

https://github.com/cvxgrp/diffcp/blob/master/cpp/src/lsqr.cpp#L49
``` cpp
  if (iter_lim == -1) {
    iter_lim = 2 * A.n;
  } else {
    iter_lim *= A.n;
  }
  std::cout << "iter_lim: " << iter_lim << std::endl;
```


Below is the output:

```
loss -8.689 weight 5.100 grad -0.086
loss -8.696 weight 5.200 grad -0.085
loss -8.694 weight 5.233 grad 0.074
loss -8.690 weight 5.211 grad 0.141
loss -8.693 weight 5.174 grad 0.074
loss -8.696 weight 5.128 grad 0.074
loss -8.697 weight 5.091 grad -0.018
loss -8.695 weight 5.078 grad -0.085
loss -8.694 weight 5.084 grad -0.085
loss -8.695 weight 5.104 grad -0.085
loss -8.696 weight 5.136 grad -0.085
```
which matches that of dense mode.

**Comment by sbarratt** [comment]
Dichao, thanks for debugging this. We probably should add a pass-through
for arguments to lsqr, and possibly increase the max iterations.
